compiler/rustc_hir_typeck/src/lib.rs

```diff
@@ -206,7 +207,18 @@ fn typeck_with_fallback<'tcx>(
     });
     let body = tcx.hir().body(body_id);
 
-    let typeck_results = Inherited::build(tcx, def_id).enter(|inh| {
+    let fn_sig_infer = fn_sig.map_or(false, |fn_sig| {
+        rustc_hir_analysis::collect::get_infer_ret_ty(&fn_sig.decl.output).is_some()
+    });
+
+    let mk_defining_use_anchor = |def_id| {
+        // In case we are inferring the return signature (via `_` types), ignore defining use
+        // rules, as we'll error out anyway. This helps improve diagnostics, which otherwise
+        // may just see `ty::Error` instead of `ty::Alias(Opaque, _)` and not produce better errors.
+        if fn_sig_infer { DefiningAnchor::Bubble } else { DefiningAnchor::Bind(def_id) }
```

Comment:
Which are the diagnostics changed by this?

Reply:
An example is

```rust
type Pointer<T> = impl std::ops::Deref<Target=T>;

fn test() -> Pointer<_> {
    //~^ ERROR: the placeholder `_` is not allowed within types
    Box::new(1)
}
```

which produces the following error:

```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
  --> $DIR/issue-77179.rs:7:22
   |
LL | fn test() -> Pointer<_> {
   |              --------^-
   |              |       |
   |              |       not allowed in type signatures
   |              help: replace with the correct return type: `Pointer<i32>`
```

but without the change it would not see `Pointer<i32>` but just `TyKind::Error` and thus not report any suggestion, even if the main error would still be emitted

Reply:
isn't using `Pointer<i32>` forbidden here anyways, it currently results in the following (bad) error: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=79a93af26aa63cadbcd4380373e40585)
```rust
error[E0792]: expected generic type parameter, found `i32`
 --> src/lib.rs:5:5
  |
2 | type Pointer<T> = impl std::ops::Deref<Target = T>;
  |              - this generic parameter must be used with a generic type parameter
...
5 |     Box::new(1)
  |     ^^^^^^^^^^^
```
i personally feel like this is not worth it as it imo makes the code noticeably more confusing for a fairly small benefit.